From b75cfe4eab2f2744463c5333dfef7aab6abb22ef Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:36:37 -0400
Subject: [ethernet] i40e: Bug fix for FDIR replay logic

Message-id: <1409658042-22689-37-git-send-email-sassmann@redhat.com>
Patchwork-id: 90230
O-Subject: [RHEL7.1 PATCH 036/281] i40e: Bug fix for FDIR replay logic
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Anjali Singhai Jain <anjali.singhai@intel.com>

The FDIR replay logic was being run a little too soon (before the
queues were enabled) and hence the tail bump was not effective till
a later transaction happened on the queue.

Change-ID: Icfd7cd2e79fc3cae3cbd3f703a2b3a148b4e7bf6
Signed-off-by: Anjali Singhai Jain <anjali.singhai@intel.com>
Signed-off-by: Catherine Sullivan <catherine.sullivan@intel.com>
Tested-by: Kavindya Deegala <kavindya.s.deegala@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit ca64fa4e7eda5d9e2b5f424e901983b86ba0fc49)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_main.c b/drivers/net/ethernet/intel/i40e/i40e_main.c
index 6185856..669715b 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_main.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_main.c
@@ -2456,8 +2456,6 @@ static int i40e_vsi_configure(struct i40e_vsi *vsi)
  i40e_set_vsi_rx_mode(vsi);
  i40e_restore_vlan(vsi);
  i40e_vsi_config_dcb_rings(vsi);
- if (vsi->type == I40E_VSI_FDIR)
-  i40e_fdir_filter_restore(vsi);
  err = i40e_vsi_configure_tx(vsi);
  if (!err)
   err = i40e_vsi_configure_rx(vsi);
@@ -4088,6 +4086,10 @@ static int i40e_up_complete(struct i40e_vsi *vsi)
  } else if (vsi->netdev) {
   netdev_info(vsi->netdev, "NIC Link is Down\n");
  }
+
+ /* replay FDIR SB filters */
+ if (vsi->type == I40E_VSI_FDIR)
+  i40e_fdir_filter_restore(vsi);
  i40e_service_event_schedule(pf);
 
  return 0;
-- 
1.7.1